src/main/java/net/neoforged/neoforge/registries/BaseMappedRegistry.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.registries;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.neoforged.neoforge.registries.callback.AddCallback;
+import net.neoforged.neoforge.registries.callback.BakeCallback;
+import net.neoforged.neoforge.registries.callback.ClearCallback;
+import net.neoforged.neoforge.registries.callback.RegistryCallback;
+import org.jetbrains.annotations.ApiStatus;
+
+@ApiStatus.Internal
+public abstract class BaseMappedRegistry<T> implements Registry<T> {
+    protected final List<AddCallback<T>> addCallbacks = new ArrayList<>();
+    protected final List<BakeCallback<T>> bakeCallbacks = new ArrayList<>();
+    protected final List<ClearCallback<T>> clearCallbacks = new ArrayList<>();
+    final Map<ResourceLocation, ResourceLocation> aliases = new HashMap<>();
+    private int maxId = Integer.MAX_VALUE - 1;
+    private boolean sync;
+
+    void setSync(boolean sync) {
+        this.sync = sync;
+    }
+
+    @Override
+    public boolean doesSync() {
+        return this.sync;
+    }
+
+    void setMaxId(int maxId) {
+        this.maxId = maxId;
+    }
+
+    @Override
+    public int getMaxId() {
+        return this.maxId;
+    }
+
+    @Override
+    public void addCallback(RegistryCallback<T> callback) {
+        if (callback instanceof AddCallback<T> addCallback)
+            this.addCallbacks.add(addCallback);
+        if (callback instanceof BakeCallback<T> bakeCallback)
+            this.bakeCallbacks.add(bakeCallback);
+        if (callback instanceof ClearCallback<T> clearCallback)
+            this.clearCallbacks.add(clearCallback);
+    }
+
+    @Override
+    public void addAlias(ResourceLocation from, ResourceLocation to) {
+        if (from.equals(to))
+            return;
+        if (this.aliases.containsKey(from)) {
+            ResourceLocation old = this.aliases.get(from);
+            if (!old.equals(to))
```

Comment:
Since aliases are (or were?) persisted to disk, I think this should be allowed, in case the target of an alias changes for whatever reason. 

Otherwise this should just error even if `old == to` because we should be preventing duplicate registration.

Reply:
Aliases are not persisted to disk anymore (and I have no clue if they were before).

Reply:
The question is if we want to store a list of known entries to disk should that include aliases or not